lib/twocohom.gi

```diff
@@ -667,6 +667,168 @@ function( G, M )
     return TwoCoboundariesSQ( C, G, M );
 end );
 
+# non-solvable cohomology based on rewriting
+
+# return isomorphism G-fp and fp->mon, such that presentation of monoid is
+# confluent (wrt wreath order). Returns list [fphom,monhom,ordering]
+BindGlobal("ConfluentMonoidPresentationForGroup",function(G)
+local iso,fp,n,dec,homs,mos,i,j,ffp,imo,m,k,gens,fm,mgens,rules,
+      loff,off,monreps,left,right,fmgens,r,diff,monreal,nums,reduce,hom,dept;
+  iso:=IsomorphismFpGroupByChiefSeries(G:rewrite);
+  fp:=Range(iso);
+  gens:=GeneratorsOfGroup(fp);
+  n:=Length(gens);
+  dec:=iso!.decompinfo;
+
+  fmgens:=[];
+  mgens:=[];
+  for i in gens do
+    Add(fmgens,i);
+    Add(fmgens,i^-1);
+    Add(mgens,String(i));
+    Add(mgens,String(i^-1));
+  od;
+  nums:=List(fmgens,x->LetterRepAssocWord(UnderlyingElement(x))[1]);
+  fm:=FreeMonoid(mgens);
+  mgens:=GeneratorsOfMonoid(fm);
+  rules:=[];
+  reduce:=function(w)
+  local red,i,p;
+    w:=LetterRepAssocWord(w);
+    repeat
+      i:=1;
+      red:=false;
+      while i<=Length(rules) and red=false do
+        p:=PositionSublist(w,LetterRepAssocWord(rules[i][1]));
+        if p<>fail then
+          #Print("Apply ",rules[i],p,w,"\n");
+          w:=Concatenation(w{[1..p-1]},LetterRepAssocWord(rules[i][2]),
+            w{[p+Length(rules[i][1])..Length(w)]});
+          red:=true;
+        else
+          i:=i+1;
+        fi;
+      od;
+    until red=false;
+    return AssocWordByLetterRep(FamilyObj(One(fm)),w);
+  end;
+
+
+  homs:=ShallowCopy(dec.homs);
+  mos:=[];
+  off:=Length(mgens);
+  dept:=[];
+  # go up so we may reduce tails
+  for i in [Length(homs),Length(homs)-1..1] do
+    Add(dept,off);
+    if IsPcgs(homs[i]) then
+      ffp:=AbelianGroup(IsFpGroup,RelativeOrders(homs[i]));
+    else
+      ffp:=Range(dec.homs[i]);
+    fi;
+    imo:=IsomorphismFpMonoid(ffp);
+    Add(mos,imo);
+    m:=Range(imo);
+    loff:=off-Length(GeneratorsOfMonoid(m));
+    monreps:=fmgens{[loff+1..off]};
+    monreal:=mgens{[loff+1..off]};
+    if IsBound(m!.rewritingSystem) then
+      k:=m!.rewritingSystem;
```

Comment:
This is one of only two lines in this function which is no covered. Very nice!

Reply:
Yes, this is as a hook for future extensions (RWS for simple groups).